File: ios/MiseEnPlace/App/DataSeeder.swift
import Foundation
import SwiftData

struct MealsExport: Codable {
    let meals: [MealJSON]
    let similarities: [SimilarityJSON]
}

struct MealJSON: Codable {
    let id: String
    let name: String
    let description: String
    let cuisine: String
    let mealType: String
    let difficulty: String
    let prepTime: Int
    let cookTime: Int
    let totalTime: Int
    let servings: Int
    let costPerServing: Double
    let totalCost: Double
    let sourceInspiration: String
    let imageName: String?
    let dietaryTags: [String]
    let seasons: [String]
    let equipment: [String]
    let tips: [String]
    let variations: [String]
    let ingredients: [IngredientJSON]
    let steps: [StepJSON]
}

struct IngredientJSON: Codable {
    let name: String
    let category: String
    let quantity: Double
    let unit: String
    let isOptional: Bool
    let prepNote: String?
    let cost: Double?
}

struct StepJSON: Codable {
    let order: Int
    let instruction: String
    let duration: Int?
}

struct SimilarityJSON: Codable {
    let mealAId: String
    let mealBId: String
    let sharedCount: Int
    let overlapRatio: Double
    let sharedIngredients: [String]
}

enum DataSeeder {
    static func seedIfNeeded(context: ModelContext) {
        let count = (try? context.fetchCount(FetchDescriptor<Meal>())) ?? 0
        guard count == 0 else { return }

        guard let url = Bundle.main.url(forResource: "meals", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let export = try? JSONDecoder().decode(MealsExport.self, from: data)
        else {
            print("[DataSeeder] Failed to load meals.json")
            return
        }

        for mealJSON in export.meals {
            let meal = Meal(
                id: mealJSON.id,
                name: mealJSON.name,
                desc: mealJSON.description,
                cuisine: mealJSON.cuisine,
                mealType: mealJSON.mealType,
                difficulty: mealJSON.difficulty,
                prepTime: mealJSON.prepTime,
                cookTime: mealJSON.cookTime,
                totalTime: mealJSON.totalTime,
                servings: mealJSON.servings,
                costPerServing: mealJSON.costPerServing,
                totalCost: mealJSON.totalCost,
                sourceInspiration: mealJSON.sourceInspiration,
                imageName: mealJSON.imageName,
                dietaryTags: mealJSON.dietaryTags,
                seasons: mealJSON.seasons,
                equipment: mealJSON.equipment,
                tips: mealJSON.tips,
                variations: mealJSON.variations
            )
            context.insert(meal)

            for ing in mealJSON.ingredients {
                let ingredient = MealIngredient(
                    name: ing.name, category: ing.category,
                    quantity: ing.quantity, unit: ing.unit,
                    isOptional: ing.isOptional, prepNote: ing.prepNote,
                    cost: ing.cost
                )
                ingredient.meal = meal
                context.insert(ingredient)
            }

            for step in mealJSON.steps {
                let recipeStep = RecipeStep(
                    order: step.order, instruction: step.instruction,
                    duration: step.duration
                )
                recipeStep.meal = meal
                context.insert(recipeStep)
            }
        }

        for sim in export.similarities {
            let similarity = MealSimilarity(
                mealAId: sim.mealAId, mealBId: sim.mealBId,
                sharedCount: sim.sharedCount, overlapRatio: sim.overlapRatio,
                sharedIngredients: sim.sharedIngredients
            )
            context.insert(similarity)
        }

        try? context.save()
        print("[DataSeeder] Seeded \(export.meals.count) meals, \(export.similarities.count) similarities")
    }
}

File: android/app/src/main/java/com/ahmedkhaledmohamed/potluck/data/DataSeeder.kt
package com.ahmedkhaledmohamed.potluck.data

import android.content.Context
import com.ahmedkhaledmohamed.potluck.data.db.*
import com.google.gson.Gson
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DataSeeder @Inject constructor(
    @ApplicationContext private val context: Context,
    private val db: AppDatabase
) {
    suspend fun seedIfNeeded() {
        val prefs = context.getSharedPreferences("potluck", Context.MODE_PRIVATE)
        val version = prefs.getInt("dataVersion", 0)
        if (version >= 2 && db.mealDao().count() > 0) return

        val json = context.assets.open("meals.json").bufferedReader().readText()
        val export = Gson().fromJson(json, MealsExport::class.java)

        db.clearAllTables()

        val meals = export.meals.map { m ->
            MealEntity(
                id = m.id, name = m.name, description = m.description,
                cuisine = m.cuisine, mealType = m.mealType, difficulty = m.difficulty,
                prepTime = m.prepTime, cookTime = m.cookTime, totalTime = m.totalTime,
                servings = m.servings, costPerServing = m.costPerServing, totalCost = m.totalCost,
                sourceInspiration = m.sourceInspiration, imageUrl = m.imageUrl,
                dietaryTags = m.dietaryTags, seasons = m.seasons, equipment = m.equipment,
                tips = m.tips, variations = m.variations
            )
        }
        db.mealDao().insertMeals(meals)

        val ingredients = export.meals.flatMap { m ->
            m.ingredients.map { ing ->
                MealIngredientEntity(
                    mealId = m.id, name = ing.name, category = ing.category,
                    quantity = ing.quantity, unit = ing.unit, isOptional = ing.isOptional,
                    prepNote = ing.prepNote, cost = ing.cost
                )
            }
        }
        db.mealDao().insertIngredients(ingredients)

        val steps = export.meals.flatMap { m ->
            m.steps.map { s ->
                RecipeStepEntity(mealId = m.id, order = s.order, instruction = s.instruction, duration = s.duration)
            }
        }
        db.mealDao().insertSteps(steps)

        val similarities = export.similarities.map { s ->
            MealSimilarityEntity(
                mealAId = s.mealAId, mealBId = s.mealBId,
                sharedCount = s.sharedCount, overlapRatio = s.overlapRatio,
                sharedIngredients = s.sharedIngredients
            )
        }
        db.similarityDao().insertAll(similarities)

        prefs.edit().putInt("dataVersion", 2).apply()
    }
}
